### PR TITLE
host-gw: add windows support

### PR DESCRIPTION
**Pull request by rakelkar**
## Description
Host GW support for windows (3a from #832) #833 
Needs integration tests!!

## Todos
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by rakelkar** [comment]
@tomdee I am having trouble updating just the gonetsh vendor version. What is the correct way to do this - issues online appear to suggest editing the lock file manually. Do you have thoughts on how to do? I want to update to https://github.com/rakelkar/gonetsh/commit/758b1f7c9d1ca5acaca0b909f06ff9ed9fe35de1

**Comment by tomdee** [comment]
@rakelkar You can just update the lock file manually. I also don't mind if you end up updating other vendored libraries. If I think it's a problem to update other libraries then 'll introduce more pins in glide.yaml.

**Comment by rocky0001** [comment]
**got the error: HNS failed with error : An adapter was not found.**

I0711 13:57:07.141605    4872 main.go:422] Searching for interface using 10.126.79.218
2018-07-11 13:57:07.182620 I | Interfaces : [{0 vEthernet (Ethernet 2) 25 true 10.126.79.218 24 0 10.126.79.1} {0 Loopback Pseudo-Interface 1 75 false 127.0.0.1 8 0 } {0 vEthernet (nat) 5000 true 172.29.160.1 20 0 }]
2018-07-11 13:57:07.219603 I | Found Interface for 10.126.79.218 => {15 vEthernet (Ethernet 2) 25 true 10.126.79.218 24 0 10.126.79.1}
I0711 13:57:07.221603    4872 main.go:499] Using interface with name vEthernet (Ethernet 2) and address 10.126.79.218
I0711 13:57:07.234575    4872 main.go:516] Defaulting external address to interface address (10.126.79.218)
I0711 13:57:07.237603    4872 kube.go:131] Waiting 10m0s for node controller to sync
I0711 13:57:07.238604    4872 kube.go:294] Starting kube subnet manager
I0711 13:57:08.239596    4872 kube.go:138] Node controller sync successful
I0711 13:57:08.241573    4872 main.go:240] Created subnet manager: Kubernetes Subnet Manager - ip-10-126-79-218.us-west-2.compute.internal
I0711 13:57:08.241573    4872 main.go:243] Installing signal handlers
I0711 13:57:08.241573    4872 main.go:358] Found network config - Backend type: host-gw
I0711 13:57:08.257572    4872 hostgw_windows.go:151] Attempting to create HNS network, request: {"DNSServerList":"","Name":"cbr0","Subnets":[{"AddressPrefix":"192.168.3.0/24","GatewayAddress":"192.168.3.1"}],"Type":"l2bridge"}
E0711 13:57:08.317554    4872 main.go:285] Error registering network: unable to create network [cbr0], error: HNS failed with error : An adapter was not found.
I0711 13:57:08.320573    4872 main.go:338] Stopping shutdownHandler...

**Comment by pablodav** [comment]
What's the status of this possible support? would be great to have support for flannel plugin on windows.

Same for #922

**Comment by daschott** [comment]
@pablodav Flannel in host-gw mode works today on Windows with pre-built binaries that include the open PR's, but it's not an ideal situation. Right now we are targeting [this PR](https://github.com/containernetworking/plugins/pull/85
) and working together with the maintainers to try and get basic Windows CNI plugins in which can be used by Flannel. Afterwards we will cleanup these plugins and work on tests with @madhanrm and @dineshgovindasamy as maintainers (this is a Windows only plugin, so it shouldn't break any existing CNI's for Linux). VxLAN is still under active development, including from platform perspective.

**Comment by pablodav** [comment]
Thaks for the update @daschott !  Sounds good to have some news and clear information where it is going!

**Comment by madhanrm** [comment]
@Tomdee, there also seems to be another PR : https://github.com/coreos/flannel/pull/1042 which is doing similar to this PR. Can we get this PR merged & then work on top if it. The binaries out of this PR has been tested for over an year now

File: backend/hostgw/hostgw_network_windows.go [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+func (n *network) Lease() *subnet.Lease {
+	return n.lease
+}
+
+func (n *network) MTU() int {
+	return n.extIface.Iface.MTU
+}
+
+func (n *network) Run(ctx context.Context) {
```
Comment (rajatchopra) [inline backend/hostgw/hostgw_network_windows.go]: This function should be made common between linux/windows. The list of routes malloc can be part of the New/Register function itself.

File: backend/hostgw/hostgw_network_windows.go [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+	"github.com/coreos/flannel/subnet"
+
+	netroute "github.com/rakelkar/gonetsh/netroute"
+)
+
+type network struct {
+	name      string
+	extIface  *backend.ExternalInterface
+	linkIndex int
+	rl        []netroute.Route
```
Comment (rajatchopra) [inline backend/hostgw/hostgw_network_windows.go]: A better name here possibly? 'routes'?

Comment (dineshgovindasamy) [reply backend/hostgw/hostgw_network_windows.go]: Will fix this in a subsequent PR.

File: backend/hostgw/hostgw_network_windows.go (hunk 1 above)
Comment (dineshgovindasamy) [reply backend/hostgw/hostgw_network_windows.go]: we will move it in the next PR.

**Comment by astrieanna** [comment]
I have a flannel/hostgw cluster on vsphere, with a linux node and a windows node. The one thing that doesn't work is that the linux node (the host, not a container) can't connect to windows service clusterips. This breaks accessing windows services through the NodePort on the linux node.

Does accessing a windows service via a nodeport on a linux node work for you with this PR?

**Comment by dineshgovindasamy** [comment]
Lets abandon this PR. We are taking PR#1042 which is merged PR.

@rakelkar , can u abandon this PR?

**Comment by rajatchopra** [comment]
Closing this PR in favour of #1042. Please re-open if needed.

**Pull request closed without merging**